homeassistant/components/mqtt/notify.py

```diff
@@ -295,57 +201,52 @@ def device_id(self) -> str | None:
         return self._device_id
 
     @property
-    def config_entry(self) -> ConfigEntry | None:
-        """Return the config_entry."""
-        return self._config_entry
-
-    @property
-    def discovery_hash(self) -> tuple | None:
-        """Return the discovery hash."""
-        return self._discovery_hash
-
-    @property
-    def service_name(self) -> str:
-        """Return the service ma,e."""
-        return self._service_name
+    def targets(self) -> dict[str, str]:
+        """Return a dictionary of registered targets."""
+        return {target: target for target in self._config[CONF_TARGETS]}
 
-    async def async_update_service(
+    async def async_discovery_update(
         self,
         discovery_payload: DiscoveryInfoType,
     ) -> None:
         """Update the notify service through auto discovery."""
         config: MqttNotificationConfig = DISCOVERY_SCHEMA(discovery_payload)
-        # Do not rename a service if that service_name is already in use
+        # Fail if the new service name is already in use
         if (
             new_service_name := slugify(config[CONF_NAME])
         ) != self._service_name and _check_notify_service_name(
             self.hass, config
         ) is None:
             return
-        # Only refresh services if service name or targets have changes
-        if (
-            new_service_name != self._service_name
+        # We need to reregister the services
+        reintitialize = (
+            not self.hass.services.has_service(notify.DOMAIN, self._service_name)
             or config[CONF_TARGETS] != self._config[CONF_TARGETS]
-        ):
+        )
+        rename_only = new_service_name != self._service_name
+        if rename_only or reintitialize:
             services = self.hass.data[MQTT_NOTIFY_SERVICES_SETUP]
             await self.async_unregister_services()
-            if self._service_name in services:
+            if rename_only and self._service_name in services:
                 del services[self._service_name]
             self._config = config
-            self._service_name = new_service_name
-            await self.async_register_services()
+            await self.async_setup(self.hass, new_service_name, new_service_name)
```

Comment:
Any hint on how to do that?

Reply:
Send the `MQTT_DISCOVERY_NEW` signal for the notify domain with `async_dispatcher_send`.

Reply:
I have tried this, but the problem is the garbage collection. the notify platform removes the services but not the discovery logic. So it seem better to keep this in memory to be reused as soon as a new discovery payload arrives.
My suggestion is leave it as it is now.

Reply:
What discovery logic are you referring to?

Reply:
Does it not work to simply remove the services and then register new ones with updated names?

Reply:
It makes reloading complicated, removing the service happens during a reloading but does not remove the discovery logic for discovered services.

Reply:
If have done a lot of testing including  the MQTT_DISCOVERY_NEW with async_dispatcher_send, but the problem remains that after a reload the discovery listener is not cleaned up. Also `hass.data[MQTT_NOTIFY_SERVICES_SETUP]` is not reset. I can remove that, but that it will be difficult to log a service with a duplicate name if setup through configuration.yaml. `hass.services.has_service(notify.DOMAIN, self._service_name)` will stay False until all entries have been setup. So I can not rely on that. Futher there is no hook I can use.
Conclusion is still that this is not becoming any better. Any hints left?

Reply:
Which discovery listener are we talking about?

Why should we register more than one service for the integration? Why not use the target parameter in the service call to allow the user to target known devices?

Reply:
> Which discovery listener are we talking about?

https://github.com/home-assistant/core/pull/67966/files#diff-0d67e0c95bc8b8efc20c8d5491b8ed7a6634598a9fcdc954f96a1077a2c77e7aR612-R616

Reply:
> Why should we register more than one service for the integration? Why not use the target parameter in the service call to allow the user to target known devices?

Because MQTT devices can have multiple targets. Further through discovery device can advertise a notify service, e.g. an LCD screen capable of showing a message.

Reply:
How the targets are resolved is up to the integration. I don't see that more than one service is required regardless of number of targets per device. The device will need some way of resolving the targets so we'll use the same way in the service.

The discovery advertises a target. It doesn't need to create a new service. I think it will be a lot easier if we only register one service.

If we only have one service we can disconnect discovery dispatch listeners when unloading the config entry, and when registering the service again if the service has already been registered earlier.

Reply:
Right, but that all is not the intention of this PR. This one is about code quality for the existing code.
There still will be a relation between discovery and registering targets with new discovery object.
I think it is better to add some support to the base notify service for a hook to the platform to cleanup some code. If that is available we can further simplify the update mechanism.

Reply:
It's not only about code quality. The original PR for mqtt notify implemented the interfaces incorrectly. I think a major reason for that is this overcomplication about registering multiple services that need clean up.

Alternatively I suggest we revert the PR and start over.

Reply:
Logic consolidated. One bug solved in the notify platform setup.